workflow/rules/mapping.smk

```diff
@@ -214,3 +214,126 @@ rule apply_bqsr:
+# use -f 2 to keep only properly paired reads, the mate of a read that's on a nonprimary chromosome is problematic for AddOrReplaceReadGroups, because we remove
+# all the other nonprimary chromosome from the header.
+
+
+rule keep_only_primary_chr:
+    input:
+        "results/mapped/vg/{sample}_sorted.bam",
+        "results/mapped/vg/{sample}_sorted.bai",
+    output:
+        bam="results/mapped/vg/{sample}_extracted.bam",
+        idx="results/mapped/vg/{sample}_extracted.bai",
+    log:
+        "logs/samtools_view_primary_chr/{sample}.log",
+    benchmark:
+        "benchmarks/samtools_view_primary_chr/{sample}.tsv"
+    params:
+        region="GRCh38.chr1 GRCh38.chr2 GRCh38.chr3 GRCh38.chr4 GRCh38.chr5 GRCh38.chr6 GRCh38.chr7 GRCh38.chr8 GRCh38.chr9 GRCh38.chr10 GRCh38.chr11 GRCh38.chr12 GRCh38.chr13 GRCh38.chr14 GRCh38.chr15 GRCh38.chr16 GRCh38.chr17 GRCh38.chr18 GRCh38.chr19 GRCh38.chr20 GRCh38.chr21 GRCh38.chr22 GRCh38.chrX GRCh38.chrY GRCh38.chrM",
+        extra="-f 2",
+    threads: 40
+    wrapper:
+        "v2.0.0/bio/samtools/view"
+
+
+# modify the header for chromosome names to be compatible with the reference genome that are acquired from ensembl
+# first delete all non classical chromosomes including unlocalized, unplaced and EBV chromosomes (delly complains about them being found in the header)
+# second remove GRCh38.chr and third convert M to MT (MT in pangenome reference and M in fasta sequence dict)
+# the following sed command replaces the first "M" it finds and replaces it with "MT"
+
+
+rule reheader:
+    input:
+        "results/mapped/vg/{sample}_extracted.bam",
+    output:
+        "results/mapped/vg/{sample}_reheadered.bam",
+    log:
+        "logs/samtools_reheader/{sample}.log",
+    benchmark:
+        "benchmarks/samtools_reheader/{sample}.tsv"
+    conda:
+        "../envs/samtools.yaml"
+    threads: 40
+    shell:
+        "samtools view -H {input} | sed '/random/d;/chrUn/d;/EBV/d;s/GRCh38.chr//g;0,/M/s//MT/' | samtools reheader - {input} > {output} 2> {log}"
+
+
+# rule samtools_index_after_reheader:
+#     input:
+#         "results/vg_mapped/{sample}_reheadered.bam",
+#     output:
+#         "results/vg_mapped/{sample}_reheadered.bai",
+#     log:
+#         "logs/samtools_index_after_reheader/{sample}.log",
+#     benchmark:
+#         "benchmarks/samtools_index_after_reheader/{sample}.tsv"
+#     threads: 40
+#     wrapper:
+#         "v2.3.2/bio/samtools/index"
+
+# adding read groups is necessary because base recalibration throws errors
+# for not being able to find read group information
+
+
+rule add_rg:
+    input:
+        "results/mapped/vg/{sample}_reheadered.bam",
+    output:
+        "results/mapped/vg/{sample}.bam",
+    log:
+        "logs/picard/add_rg/{sample}.log",
+    params:
+        extra="--RGLB lib1 --RGPL illumina --RGPU {sample} --RGSM {sample}",
+    resources:
+        mem_mb=60000,
+    wrapper:
+        "v2.3.2/bio/picard/addorreplacereadgroups"
```

Comment:
Are you sure that it is impossible to define the readgroup directly when calling vg giraffe?

Reply:
With vg, it is possible to define read groups for sample names but not for platforms. BaseRecalibrator later complains about it: 
```
org.broadinstitute.hellbender.exceptions.UserException$MalformedRead: Read SRR702070.29886333 1:13000-13100 is malformed: The input .bam file contains reads with no platform information. First observed at read with name = SRR702070.29886333
```